app/ide-desktop/lib/profiling/src/profiling.ts

```diff
@@ -0,0 +1,322 @@
+    end() {
+        this.log_level.end(this.interval_name)
+        this.released_status.released = true
+    }
+
+    /// Release the handle to manually call `end_interval` without emitting a warning.
+    release() {
+        this.released_status.released = true
+    }
+}
+
+/**
+ * Class that contains the information added to the `FinalizationRegistry` to decide whether to emit
+ * a warning when a `IntervalHandle` is garbage collected.
+ */
+class IntervalFinalizationRegistryEntry {
+    interval_name: string
+    release_status: ReleasedStatus
+
+    constructor(interval_name: string, release_status: ReleasedStatus) {
+        this.interval_name = interval_name
+        this.release_status = release_status
+    }
+}
+
+/** Class that contains information about the release status of an `IntervalHandle`. An
+ * `IntervalHandle` handle is released by calling `end` or `release`. If it is garbage collected
+ * without being released, we want to emit a warning. In order to be able to keep track of the
+ * release status of an `IntervalHandle` we need to keep the status boolean in a separate class. */
+class ReleasedStatus {
+    released: boolean
+
+    constructor() {
+        this.released = false
+    }
+}
+
+// We need to ts-ignore here as `FinalizationRegistry` is not recognised by TypeScript as being in
+// scope even though it is.
+// @ts-ignore
+const handle_registry = new FinalizationRegistry(heldValue => {
+    if (heldValue instanceof IntervalFinalizationRegistryEntry) {
+        if (!heldValue.release_status.released) {
+            console.warn(heldValue.interval_name + 'was dropped without a call to `measure`.')
+        }
+    }
+})
+
+// ===================
+// === Measurement ===
+// ===================
+
+/**
+ * Profiling measurement that contains the profiling information about a measured time interval.
+ */
+class Measurement {
+    readonly startTime: number
+    readonly duration: number
+    readonly logLevel: Profiler
+    readonly name: string
+
+    constructor(startTime: number, duration: number, logLevel: Profiler, name: string) {
+        this.startTime = startTime
+        this.duration = duration
+        this.logLevel = logLevel
+        this.name = name
+    }
+
+    static fromPerformanceEntry(entry: PerformanceEntry): Measurement | undefined {
+        const startTime = entry.startTime
+        const duration = entry.duration
+
+        let decodedLabel = decodeLabel(entry.name)
+        if (decodedLabel === undefined) {
+            console.warn(`Could not decode "${entry.name}" as measurement label.`)
+            return
+        }
+        const [logLevel, name] = decodedLabel
+
+        return new Measurement(startTime, duration, logLevel, name)
+    }
+
+    endTime(): number {
+        return this.startTime + this.duration
+    }
+
+    prettyString(): string {
+        return `[${this.startTime.toFixed(2)},${this.endTime().toFixed(
+            2
+        )}] (${this.duration.toFixed(2)}) ${this.name}`
```

Comment:
a lot of magical "2" here. Create a variable describing what it is please

Reply:
Added a constant that specifies the formatting precision of the numbers.